File: emoji/src/main/java/com/vanniktech/emoji/EmojiImageView.java
/*
 * Copyright (C) 2016 - Niklas Baudy, Ruben Gees, Mario Đanić and contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.vanniktech.emoji;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Point;
import android.util.AttributeSet;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import com.vanniktech.emoji.emoji.Emoji;
import com.vanniktech.emoji.listeners.OnEmojiClickListener;
import com.vanniktech.emoji.listeners.OnEmojiLongClickListener;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;

public final class EmojiImageView extends AppCompatImageView {
  private static final int VARIANT_INDICATOR_PART_AMOUNT = 6;
  private static final int VARIANT_INDICATOR_PART = 5;

  Emoji currentEmoji;

  OnEmojiClickListener clickListener;
  OnEmojiLongClickListener longClickListener;

  private final Paint variantIndicatorPaint = new Paint();
  private final Path variantIndicatorPath = new Path();

  private final Point variantIndicatorTop = new Point();
  private final Point variantIndicatorBottomRight = new Point();
  private final Point variantIndicatorBottomLeft = new Point();

  private boolean hasVariants;

  @Nullable private Future<?> future;

  public EmojiImageView(final Context context, final AttributeSet attrs) {
    super(context, attrs);

    variantIndicatorPaint.setStyle(Paint.Style.FILL);
    variantIndicatorPaint.setAntiAlias(true);
  }

  @Override public void onMeasure(final int widthMeasureSpec, final int heightMeasureSpec) {
    super.onMeasure(widthMeasureSpec, heightMeasureSpec);

    final int measuredWidth = getMeasuredWidth();
    //noinspection SuspiciousNameCombination
    setMeasuredDimension(measuredWidth, measuredWidth);
  }

  @Override protected void onSizeChanged(final int w, final int h, final int oldw, final int oldh) {
    super.onSizeChanged(w, h, oldw, oldh);

    variantIndicatorTop.x = w;
    variantIndicatorTop.y = h / VARIANT_INDICATOR_PART_AMOUNT * VARIANT_INDICATOR_PART;
    variantIndicatorBottomRight.x = w;
    variantIndicatorBottomRight.y = h;
    variantIndicatorBottomLeft.x = w / VARIANT_INDICATOR_PART_AMOUNT * VARIANT_INDICATOR_PART;
    variantIndicatorBottomLeft.y = h;

    variantIndicatorPath.rewind();
    variantIndicatorPath.moveTo(variantIndicatorTop.x, variantIndicatorTop.y);
    variantIndicatorPath.lineTo(variantIndicatorBottomRight.x, variantIndicatorBottomRight.y);
    variantIndicatorPath.lineTo(variantIndicatorBottomLeft.x, variantIndicatorBottomLeft.y);
    variantIndicatorPath.close();
  }

  @Override protected void onDraw(final Canvas canvas) {
    super.onDraw(canvas);

    if (hasVariants && getDrawable() != null) {
      canvas.drawPath(variantIndicatorPath, variantIndicatorPaint);
    }
  }

  @Override protected void onDetachedFromWindow() {
    super.onDetachedFromWindow();

    cancelFuture();
  }

  private void cancelFuture() {
    if (future != null) {
      future.cancel(true);
    }

    future = null;
  }

  void setEmoji(@NonNull final ExecutorService executorService, @NonNull final EmojiTheming theming, @NonNull final Emoji emoji) {
    final Context context = getContext();
    variantIndicatorPaint.setColor(EmojiThemings.dividerColor(theming, context));
    postInvalidate();

    if (!emoji.equals(currentEmoji)) {
      setImageDrawable(null);

      currentEmoji = emoji;
      hasVariants = emoji.getBase().hasVariants();

      cancelFuture();

      setOnClickListener(view -> {
        if (clickListener != null) {
          clickListener.onEmojiClick(currentEmoji);
        }
      });

      setOnLongClickListener(hasVariants ? (OnLongClickListener) view -> {
        longClickListener.onEmojiLongClick(this, currentEmoji);
        return true;
      } : null);

      future = executorService.submit(() -> {
          setImageDrawable(emoji.getDrawable(context));
      });
    }
  }

  /**
   * Updates the emoji image directly. This should be called only for updating the variant
   * displayed (of the same base emoji), since it does not run asynchronously and does not update
   * the internal listeners.
   *
   * @param emoji The new emoji variant to show.
   */
  public void updateEmoji(@NonNull final Emoji emoji) {
    if (!emoji.equals(currentEmoji)) {
      currentEmoji = emoji;

      setImageDrawable(emoji.getDrawable(getContext()));
    }
  }

  void setOnEmojiClickListener(@Nullable final OnEmojiClickListener listener) {
    clickListener = listener;
  }

  void setOnEmojiLongClickListener(@Nullable final OnEmojiLongClickListener listener) {
    longClickListener = listener;
  }
}

File: emoji/src/main/kotlin/com/vanniktech/emoji/EmojiImageView.kt
/*
 * Copyright (C) 2016 - Niklas Baudy, Ruben Gees, Mario Đanić and contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.vanniktech.emoji

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.Point
import android.util.AttributeSet
import android.view.View.OnLongClickListener
import androidx.appcompat.widget.AppCompatImageView
import com.vanniktech.emoji.emoji.Emoji
import com.vanniktech.emoji.listeners.OnEmojiClickListener
import com.vanniktech.emoji.listeners.OnEmojiLongClickListener

class EmojiImageView @JvmOverloads constructor(
  context: Context,
  attrs: AttributeSet? = null,
) : AppCompatImageView(context, attrs) {
  private var currentEmoji: Emoji? = null
  internal var clickListener: OnEmojiClickListener? = null
  internal var longClickListener: OnEmojiLongClickListener? = null
  private val variantIndicatorPaint = Paint().apply {
    style = Paint.Style.FILL
    isAntiAlias = true
  }
  private val variantIndicatorPath = Path()
  private val variantIndicatorTop = Point()
  private val variantIndicatorBottomRight = Point()
  private val variantIndicatorBottomLeft = Point()
  private var hasVariants = false
  private var imageLoadingTask: ImageLoadingTask? = null

  public override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
    super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    val measuredWidth = measuredWidth
    setMeasuredDimension(measuredWidth, measuredWidth)
  }

  override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
    super.onSizeChanged(w, h, oldw, oldh)
    variantIndicatorTop.x = w
    variantIndicatorTop.y = h / VARIANT_INDICATOR_PART_AMOUNT * VARIANT_INDICATOR_PART
    variantIndicatorBottomRight.x = w
    variantIndicatorBottomRight.y = h
    variantIndicatorBottomLeft.x = w / VARIANT_INDICATOR_PART_AMOUNT * VARIANT_INDICATOR_PART
    variantIndicatorBottomLeft.y = h
    variantIndicatorPath.rewind()
    variantIndicatorPath.moveTo(variantIndicatorTop.x.toFloat(), variantIndicatorTop.y.toFloat())
    variantIndicatorPath.lineTo(variantIndicatorBottomRight.x.toFloat(), variantIndicatorBottomRight.y.toFloat())
    variantIndicatorPath.lineTo(variantIndicatorBottomLeft.x.toFloat(), variantIndicatorBottomLeft.y.toFloat())
    variantIndicatorPath.close()
  }

  override fun onDraw(canvas: Canvas) {
    super.onDraw(canvas)
    if (hasVariants && drawable != null) {
      canvas.drawPath(variantIndicatorPath, variantIndicatorPaint)
    }
  }

  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    cancelFuture()
  }

  private fun cancelFuture() {
    imageLoadingTask?.cancel()
    imageLoadingTask = null
  }

  fun setEmoji(theming: EmojiTheming, emoji: Emoji, variantEmoji: VariantEmoji?) {
    val context = context
    variantIndicatorPaint.color = theming.dividerColor(context)
    postInvalidate()

    if (emoji != currentEmoji) {
      setImageDrawable(null)
      currentEmoji = emoji
      hasVariants = emoji.base.hasVariants() && variantEmoji !is NoVariantEmoji
      cancelFuture()
      setOnClickListener {
        clickListener?.onEmojiClick(currentEmoji!!)
      }
      setOnLongClickListener(
        when {
          hasVariants -> OnLongClickListener {
            longClickListener?.onEmojiLongClick(this, emoji)
            true
          }
          else -> null
        }
      )

      imageLoadingTask = ImageLoadingTask(this)
      imageLoadingTask?.start(emoji)
    }
  }

  /**
   * Updates the emoji image directly. This should be called only for updating the variant
   * displayed (of the same base emoji), since it does not run asynchronously and does not update
   * the internal listeners.
   *
   * @param emoji The new emoji variant to show.
   */
  fun updateEmoji(emoji: Emoji) {
    if (emoji != currentEmoji) {
      currentEmoji = emoji
      setImageDrawable(emoji.getDrawable(context))
    }
  }

  internal companion object {
    private const val VARIANT_INDICATOR_PART_AMOUNT = 6
    private const val VARIANT_INDICATOR_PART = 5
  }
}
